### PR TITLE
add login and registration for user and set their products

### PR DESCRIPTION
**Pull request by Rubncal04**
I didn't use devise, only bcrypt and jwt

### CODE REVIEW

File: app/controllers/api/v1/products_controller.rb [hunk 1]
```diff
@@ -2,11 +2,13 @@ class Api::V1::ProductsController < Api::ApiController
   before_action :set_product, only: [:show, :update, :destroy]
 
   def index
-    @products = Product.all
+    if @current_user.present?
+      @products = Product.order(created_at: :desc).where user_id: @current_user.id
+    end
```
Comment (gabo-cs) [inline app/controllers/api/v1/products_controller.rb:7]: You don't need to do `Product.where`
You can do `@current_user.products.order(...)`.

File: config/credentials.yml.enc [hunk 2]
```diff
@@ -1 +1 @@
-O5c+t+XQ/iJzx7U22WfRwOHRlzk9Sps7jeL+93KQR6altE/yeetNmJ9IlFYnOnI9vBYMjJWU5ZO4h6WAJk7iR22rhP7O0TQqTN6oQFRMzJZvoEiBSfElObgJV8pgcw+wvqOqZRS4QmHYes5cTl1DWZ3Uio72k1giTss3W2yUM/0YSH4lXZ7poYWNzfyp387YPDuo/h0q7//NsWKxLbHjKRUWtIxYzNk6tt2ZBVfcupj69Stabpf+WXCuLkKDcuVRo1fbIhQzxiZ5xV7JYXCKh6omvlXukNEAZY4tj2G0gj3i/3o3JUeefvoe30s5duPdi/9yP4Ib2kMhoKJoD9p7/7X7KDRP45XDWHdYh2x+eqlx07eEBGoGnnDSqYoBjgvQooxFDAFZs/3QkaIR77/TJvXIneJCNJQjr7vL--lPhU2ORYeVoJGFNz--okKfBomPrVppFCOfYSUxsw==
\ No newline at end of file
+CVy/5EEZ7W3dm0ElqOU/FiNjk80O5mFvWoxk/R0faK5SXUFYza4s9lFTmulBXgqCVsyVQb11N5Xe3wU3h1GHTATLoLdB11f31zxDKwuyuBBx4R3vZ9p4D6AEufCe/V7NTbnjFX7tPRNLhowhJUmap05ThDMzf5EBm8xtAvAZGyV7ol5nuCAdUN5H4UwKzRH10V0mIo/bxGdVpIHLi1fnm3FTQaTHv7MSHKGyxrCyqxNxoqE4tFTnfNj8/XKH3k52NRnSckI5kD4tozktq0PjH9CxzTuqrypa3E4pWiqUBmS0psF9XCEi9mo4oiG8e6DVPadTvVyV/n4zYQ8PWrhMDiyKuGUKv+aB84gMBX6tEYhM984OhHgBFooX5WADR3fZb5ebsgzT63reGJtrV0Be8eCk7Z8J+iHS+Pa3--Pa4ROCKtETfCbG1O--MQoo5EigTmqzzJ49OLdwnw==
```
Comment (gabo-cs) [inline config/credentials.yml.enc:1]: Do we need to add this file to the git history?

File: config/routes.rb [hunk 3]
```diff
@@ -6,9 +6,16 @@
 
   namespace :api, defaults: { format: :json } do
     namespace :v1 do
+      namespace :users do
+        get '/:id', to: 'users#show'
+        post '/', to: 'users#create'
+        put '/:id', to: 'users#update'
+        # resources :users, only: [:update, :show, :create]
```
Comment (gabo-cs) [inline config/routes.rb:13]: Why not this way?

Comment (Rubncal04) [reply config/routes.rb]: resolved it

File: config/credentials.yml.enc (hunk 2 above)
Comment (Rubncal04) [reply config/credentials.yml.enc:1]: yes, I added to gitignore file

File: app/controllers/api/v1/products_controller.rb (hunk 1 above)
Comment (gabo-cs) [reply app/controllers/api/v1/products_controller.rb:7]: Actually, where is `@current_user` being set ? I couldn't find it. Can you point me to it?

File: app/controllers/api/v1/users_controller.rb [hunk 4]
```diff
@@ -0,0 +1,44 @@
+class Api::V1::UsersController < Api::ApiController
+  skip_before_action :authenticate_request, only: [ :create ]
+
+  def show
+    if @current_user
+      token = jwt_enconde(user_id: @current_user.id)
+      render json: { token: token, user: @current_user }, status: :ok
+    else
+      render json: { error: @error }, status: :not_found
+    end
```
Comment (gabo-cs) [inline app/controllers/api/v1/users_controller.rb:10]: This whole `@current_user` validation and use of `@error` will change if you make the changes I stated in my first comment.

File: app/controllers/api/v1/users_controller.rb [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+  end
+
+  def create
+    @user = User.new(user_params)
+
+    if @user.save!
```
Comment (gabo-cs) [inline app/controllers/api/v1/users_controller.rb:16]: Remember, if you're using `save!` and are rescuing the exception, then this `if` is pointless.

You either do:
1) `if object.save`+ else
or
2) `object.save!` + rescue
But not a mix of them.

File: app/controllers/api/v1/users_controller.rb [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+      token = jwt_enconde(user_id: @user.id)
+      render json: { user: @user, token: token }, status: :created
+    end
+
+    rescue => e
+      render json: { errors: @user.errors.messages }, status: :unprocessable_entity
+  end
+
+  def update
+    if current_user?() && @current_user.update!(user_params)
```
Comment (gabo-cs) [inline app/controllers/api/v1/users_controller.rb:26]: Ditto of the current_user validation and ditto of the `if` + bang method

File: config/routes.rb [hunk 7]
```diff
@@ -6,11 +6,11 @@
 
   namespace :api, defaults: { format: :json } do
     namespace :v1 do
-      resources :users, only: [:update, :show, :create], controller: 'users/users'
+      resources :users, only: [:update, :show, :create]
```
Comment (gabo-cs) [inline config/routes.rb:9]: Nice!

File: config/routes.rb [hunk 8]
```diff
@@ -6,11 +6,11 @@
...
       resources :products do
         resources :orders
       end
-      post '/auth/login', to: 'users/authentication#login'
+      post '/auth/login', to: 'authentication#login'
```
Comment (gabo-cs) [inline config/routes.rb:13]: Cool deal!

**Review by gabo-cs — COMMENTED** [review]
Left some comments

File: app/controllers/api/api_controller.rb [hunk 9]
```diff
@@ -1,2 +1,18 @@
...
+  private
+
+  def authenticate_request
+    header = request.headers['Authorization']
+    header = header.split(' ').last if header
+    @decoded = jwt_decode(header)
+    begin
+      @current_user = User.find(@decoded[:user_id])
+    rescue ActiveRecord::RecordNotFound
+      render json: {error: "Unauthorized. Plase, make sure you are logged."}, status: :unauthorized
```
Comment (gabo-cs) [inline app/controllers/api/api_controller.rb:15]: Grammar is a bit off here. It should be:
```suggestion
      render json: {error: "Unauthorized. Please, make sure you are logged in."}, status: :unauthorized
```

Although, to me, "Unauthorized" is enough

File: app/controllers/api/api_controller.rb [hunk 10]
```diff
@@ -1,2 +1,18 @@
...
+
+  before_action :authenticate_request
+
+  private
+
+  def authenticate_request
+    header = request.headers['Authorization']
+    header = header.split(' ').last if header
+    @decoded = jwt_decode(header)
+    begin
```
Comment (gabo-cs) [inline app/controllers/api/api_controller.rb:12]: Why are you restricting the "error" to be only when finding the user? What if something fails during `jwt_decode(header)` ? That should render a 401, too, don't you think?

Hint: Look for specific exceptions/errors in the jwt gem documentation. 
https://github.com/jwt/ruby-jwt#expiration-time-claim
https://github.com/jwt/ruby-jwt#x509-certificates-in-x5c-header
(Don't pay attention to their use case, just focus on the exception and the rescue they do)

File: app/controllers/api/v1/products_controller.rb [hunk 11]
```diff
@@ -2,11 +2,13 @@ class Api::V1::ProductsController < Api::ApiController
   before_action :set_product, only: [:show, :update, :destroy]
 
   def index
-    @products = Product.all
+    if @current_user.present?
+      @products = @current_user.products.order(created_at: :desc)
+    end
```
Comment (gabo-cs) [inline app/controllers/api/v1/products_controller.rb:7]: You don't need to this check. Remember that if `@current_user` isn't available, a 401 will be raised.
```suggestion
    @products = @current_user.products.order(created_at: :desc)
```

File: app/controllers/api/v1/users_controller.rb [hunk 12]
```diff
@@ -0,0 +1,37 @@
...
+    if @user.save
+      token = jwt_enconde(user_id: @user.id)
+      render json: { user: @user, token: token }, status: :created
+    else
+      render json: { errors: "Something went wrong." }, status: :unprocessable_entity
+    end
+  end
+
+  def update
+    render json: { message: "User successfully updated", user: @current_user }, status: :ok if @current_user.update!(user_params)
```
Comment (gabo-cs) [inline app/controllers/api/v1/users_controller.rb:21]: You're doing `if @object.update!`. See the comment I left you last time about when to use if with a bang method and when not to. Please let me know if you have any questions.

File: config/credentials.yml.enc (hunk 2 above)
Comment (gabo-cs) [reply config/credentials.yml.enc:1]: Why is this still showing?

File: app/controllers/api/api_controller.rb [hunk 13]
```diff
@@ -1,2 +1,20 @@
...
+
+  private
+
+  def authenticate_request
+    header = request.headers['Authorization']
+    header = header.split(' ').last if header
+    begin
+      @decoded = jwt_decode(header)
+      @current_user = User.find(@decoded[:user_id])
+    rescue JWT::DecodeError
```
Comment (gabo-cs) [inline app/controllers/api/api_controller.rb:14]: You don't need to do separate `rescue`s, you can do:
```
rescue ActiveRecord::RecordNotFound, JWT::DecodeError, JWT::ExpiredSignature
  render json: { errors: "Unauthorized." }, status: :unauthorized
end
```

(Yes, I added `ExpiredSignature` to play it safe, it was one of the links I shared :) )

File: app/controllers/api/v1/products_controller.rb (hunk 11 above)
Comment (gabo-cs) [reply app/controllers/api/v1/products_controller.rb:7]: @Rubncal04 Still missing this change

File: app/controllers/api/v1/users_controller.rb [hunk 14]
```diff
@@ -0,0 +1,38 @@
...
+    if @user.save
+      token = jwt_enconde(user_id: @user.id)
+      render json: { user: @user, token: token }, status: :created
+    else
+      render json: { errors: "Something went wrong." }, status: :unprocessable_entity
+    end
+  end
+
+  def update
+    @current_user.update!(user_params)
```
Comment (gabo-cs) [inline app/controllers/api/v1/users_controller.rb:21]: Neat!

File: app/controllers/api/v1/users_controller.rb [hunk 15]
```diff
@@ -0,0 +1,38 @@
...
+      render json: { user: @user, token: token }, status: :created
+    else
+      render json: { errors: "Something went wrong." }, status: :unprocessable_entity
+    end
+  end
+
+  def update
+    @current_user.update!(user_params) 
+    render json: { message: "User successfully updated", user: @current_user }, status: :ok
+
```
Comment (gabo-cs) [inline app/controllers/api/v1/users_controller.rb:23]: We probably don't need this line here
```suggestion
```

**Review by gabo-cs — APPROVED** [review]

**Pull request merged**